When testing this with simulation:
```
[ERROR] [1487663693.654953011, 2.190000000]: Could not load class cob_omni_drive_controller/WheelController: Failed to load library /home/fxm/projects/indigo/care-o-bot/devel/lib//libcob_omni_drive_controller.so. Make sure that you are calling the PLUGINLIB_EXPORT_CLASS macro in the library code, and that names are consistent between this macro and your XML. Error string: Could not load library (Poco exception = /home/fxm/projects/indigo/care-o-bot/devel/.private/cob_omni_drive_controller/lib/libcob_omni_drive_geom.so: undefined symbol: _ZN11PosCtrlData5resetEv)
[ERROR] [1487663693.655072828, 2.190000000]: Could not load controller 'twist_controller' because controller type 'cob_omni_drive_controller/WheelController' does not exist.
[ERROR] [1487663693.655110860, 2.190000000]: Use 'rosservice call controller_manager/list_controller_types' to get the available types
Error when loading twist_controller
```

Don't know why? 
All other repos are using `ipa320/indigo_dev`, plugins are known (`rosservice call controller_manager/list_controller_types`), parameters are set correctly...
Does the `init()` fail?

--- a/cob_omni_drive_controller/CMakeLists.txt
+++ b/cob_omni_drive_controller/CMakeLists.txt
@@ -1,7 +1,7 @@
 cmake_minimum_required(VERSION 2.8.3)
 project(cob_omni_drive_controller)
 
-find_package(catkin REQUIRED COMPONENTS angles controller_interface geometry_msgs hardware_interface message_generation nav_msgs realtime_tools sensor_msgs std_msgs std_srvs tf urdf)
+find_package(catkin REQUIRED COMPONENTS angles controller_interface dynamic_reconfigure geometry_msgs hardware_interface message_generation nav_msgs realtime_tools sensor_msgs std_msgs std_srvs tf urdf)
 
 find_package(Boost REQUIRED COMPONENTS thread)
 
@@ -13,8 +13,12 @@ add_message_files(
 
 generate_messages(DEPENDENCIES std_msgs)
 
+generate_dynamic_reconfigure_options(
+  cfg/SteerCtrl.cfg
+)
+
 catkin_package(
-  CATKIN_DEPENDS angles controller_interface geometry_msgs hardware_interface message_runtime nav_msgs sensor_msgs std_msgs std_srvs tf urdf
+  CATKIN_DEPENDS angles controller_interface dynamic_reconfigure geometry_msgs hardware_interface message_runtime nav_msgs sensor_msgs std_msgs std_srvs tf urdf
   DEPENDS Boost
   INCLUDE_DIRS include
   LIBRARIES cob_omni_drive_geom cob_omni_drive_controller

--- a/cob_omni_drive_controller/cfg/SteerCtrl.cfg
+++ b/cob_omni_drive_controller/cfg/SteerCtrl.cfg
@@ -1,0 +1,17 @@
+#!/usr/bin/env python
+
+PACKAGE = "cob_omni_drive_controller"
+
+from dynamic_reconfigure.parameter_generator_catkin import *
+
+gen = ParameterGenerator()
+
+# Parameters for tuning Impedance-Controller
+gen.add("spring",     double_t, 0, "Spring-constant (elasticity)",                              15.0,  0.0,  50.0)
+gen.add("damp",       double_t, 0, "Damping coefficient (also prop. for Velocity Feedforward)",  2.75, 0.0,  50.0)
+gen.add("virt_mass",  double_t, 0, "Virtual Mass of Spring-Damper System",                       0.1,  0.0, 200.0)
+gen.add("d_phi_max",  double_t, 0, "maximum angular velocity (cut-off)",                        10.0,  0.0,  50.0)
+gen.add("dd_phi_max", double_t, 0, "maximum angular acceleration (cut-off)",                    40.0,  0.0, 100.0)
+
+# Second arg is node name it will run in (doc purposes only), third is generated filename prefix
+exit(gen.generate(PACKAGE, "steer_control_configure", "SteerCtrl"))

--- a/cob_omni_drive_controller/include/cob_omni_drive_controller/UndercarriageCtrlGeom.h
+++ b/cob_omni_drive_controller/include/cob_omni_drive_controller/UndercarriageCtrlGeom.h
@@ -57,49 +57,85 @@
 #include <vector>
 #include <string>
 #include <stdexcept>
+#include <boost/make_shared.hpp>
+
+struct PlatformState{
+    double dVelLongMMS;
+    double dVelLatMMS;
+    double dRotRobRadS;
+
+    const double getVelX() { return dVelLongMMS / 1000.0; }
+    const double getVelY() { return dVelLatMMS / 1000.0; }
+
+    void setVelX(const double val) { dVelLongMMS  = val * 1000.0; }
+    void setVelY(const double val) { dVelLatMMS = val * 1000.0; }
+
+    PlatformState() : dVelLongMMS(0), dVelLatMMS(0),dRotRobRadS(0) {}
+};
+struct WheelState {
+    double dVelGearDriveRadS;
+    double dVelGearSteerRadS;
+    double dAngGearSteerRad;
+    WheelState() : dVelGearDriveRadS(0), dVelGearSteerRadS(0), dAngGearSteerRad(0) {}
+};
+struct WheelCommand : public WheelState {
+    double dAngGearSteerRadDelta;
+    WheelCommand() : dAngGearSteerRadDelta(0) {}
+};
+struct WheelGeom{
+    std::string steer_name, drive_name;
+
+    /** Position of the Wheels' Steering Axis'
+    *  in cartesian (X/Y) coordinates
+    *  relative to robot coordinate System
+    */
+    double dWheelXPosMM;
+    double dWheelYPosMM;
+
+    double dSteerDriveCoupling;
+
+    double dRadiusWheelMM;
+    double dDistSteerAxisToDriveWheelMM;
+
+};
+struct WheelData {
+    WheelGeom geom_;
+
+    /** Factor between steering motion and steering induced motion of drive wheels
+        *  subtract from Drive-Wheel Vel to get effective Drive Velocity (Direct Kinematics)
+        *  add to Drive-Wheel Vel (Inverse Kinematics) to account for coupling when commanding velos
+        */
+    double dFactorVel;
+
+    WheelState state_;
+
+    /** Exact Position of the Wheels' itself
+        *  in cartesian (X/Y) and polar (Dist/Ang) coordinates
+        *  relative to robot coordinate System
+        */
+    double m_dExWheelXPosMM;
+    double m_dExWheelYPosMM;
+    double m_dExWheelDistMM;
+    double m_dExWheelAngRad;
+
+    double m_dVelWheelMMS;
+
+    void updateState(const WheelState &state);
+    double getVelX() const;
+    double getVelY() const;
+
+    static double mergeRotRobRadS(const WheelData &wheel1, const WheelData &wheel2);
+
+    WheelData(const WheelGeom &geom)
+    : geom_(geom),
+        dFactorVel(-geom_.dSteerDriveCoupling + geom_.dDistSteerAxisToDriveWheelMM / geom_.dRadiusWheelMM) {
+        updateState(WheelState());
+    }
+};
 
 class UndercarriageGeomBase
 {
 public:
-    struct PlatformState{
-        double dVelLongMMS;
-        double dVelLatMMS;
-        double dRotRobRadS;
-
-        const double getVelX() { return dVelLongMMS / 1000.0; }
-        const double getVelY() { return dVelLatMMS / 1000.0; }
-
-        void setVelX(const double val) { dVelLongMMS  = val * 1000.0; }
-        void setVelY(const double val) { dVelLatMMS = val * 1000.0; }
-
-        PlatformState() : dVelLongMMS(0), dVelLatMMS(0),dRotRobRadS(0) {}
-    };
-    struct WheelState {
-        double dVelGearDriveRadS;
-        double dVelGearSteerRadS;
-        double dAngGearSteerRad;
-        WheelState() : dVelGearDriveRadS(0), dVelGearSteerRadS(0), dAngGearSteerRad(0) {}
-    };
-    struct WheelCommand : public WheelState {
-        double dAngGearSteerRadDelta;
-        WheelCommand() : dAngGearSteerRadDelta(0) {}
-    };
-    struct WheelGeom{
-        std::string steer_name, drive_name;
-
-        /** Position of the Wheels' Steering Axis'
-        *  in cartesian (X/Y) coordinates
-        *  relative to robot coordinate System
-        */
-        double dWheelXPosMM;
-        double dWheelYPosMM;
-
-        double dSteerDriveCoupling;
-
-        double dRadiusWheelMM;
-        double dDistSteerAxisToDriveWheelMM;
-
-    };
     // Get result of direct kinematics
     virtual void calcDirect(PlatformState &state) const = 0;
 
@@ -109,46 +145,12 @@ public:
     virtual ~UndercarriageGeomBase() {}
 
 protected:
-    struct WheelData {
-        WheelGeom geom_;
-
-        /** Factor between steering motion and steering induced motion of drive wheels
-            *  subtract from Drive-Wheel Vel to get effective Drive Velocity (Direct Kinematics)
-            *  add to Drive-Wheel Vel (Inverse Kinematics) to account for coupling when commanding velos
-            */
-        double dFactorVel;
-
-        WheelState state_;
-
-        /** Exact Position of the Wheels' itself
-            *  in cartesian (X/Y) and polar (Dist/Ang) coordinates
-            *  relative to robot coordinate System
-            */
-        double m_dExWheelXPosMM;
-        double m_dExWheelYPosMM;
-        double m_dExWheelDistMM;
-        double m_dExWheelAngRad;
-
-        double m_dVelWheelMMS;
-
-        void updateState(const WheelState &state);
-        double getVelX() const;
-        double getVelY() const;
-
-        static double mergeRotRobRadS(const WheelData &wheel1, const WheelData &wheel2);
-
-        WheelData(const WheelGeom &geom)
-        : geom_(geom),
-          dFactorVel(-geom_.dSteerDriveCoupling + geom_.dDistSteerAxisToDriveWheelMM / geom_.dRadiusWheelMM) {
-            updateState(WheelState());
-        }
-    };
     template<typename V> static void updateWheelStates(V& wheels, const std::vector<WheelState> &states)
     {
         if(wheels.size() != states.size()) throw std::length_error("number of states does not match number of wheels");
 
         for(size_t i = 0; i < wheels.size(); ++i){
-            wheels[i].updateState(states[i]);
+            wheels[i]->updateState(states[i]);
         }
     }
 
@@ -161,8 +163,8 @@ protected:
         // calculate rotational rate of robot and current "virtual" axis between all wheels
         for(int i = 0; i < wheels.size() ; i++)
         {
-            const WheelData &wheel = wheels[i];
-            const WheelData &other_wheel = wheels[(i+1) % wheels.size()];
+            const WheelData &wheel = *wheels[i];
+            const WheelData &other_wheel = *wheels[(i+1) % wheels.size()];
 
             dtempRotRobRADPS += WheelData::mergeRotRobRadS(wheel, other_wheel);
             dtempVelXRobMMS += wheel.getVelX();
@@ -196,83 +198,140 @@ public:
     virtual void updateWheelStates(const std::vector<WheelState> &states);
 
 private:
-    std::vector<WheelData> wheels_;
+    std::vector<boost::shared_ptr<WheelData> > wheels_;
 };
 
-class UndercarriageCtrl : public UndercarriageGeomBase {
+double limitValue(double value, double limit);
+
+template<typename T> class UndercarriageCtrlBase : public UndercarriageGeomBase {
 public:
-    struct CtrlParams{
-        double dWheelNeutralPos;
-
-        double dMaxDriveRateRadpS;
-        double dMaxSteerRateRadpS;
-
-        /** ------- Position Controller Steer Wheels -------
-        * Impedance-Ctrlr Prms
-        *  -> model Stiffness via Spring-Damper-Modell
-        *  -> only oriented at impedance-ctrl (no forces commanded)
-        *  m_dSpring   Spring-constant (elasticity)
-        *  m_dDamp             Damping coefficient (also prop. for Velocity Feedforward)
-        *  m_dVirtM    Virtual Mass of Spring-Damper System
-        *  m_dDPhiMax  maximum angular velocity (cut-off)
-        *  m_dDDPhiMax maximum angular acceleration (cut-off)
-        */
-        double dSpring, dDamp, dVirtM, dDPhiMax, dDDPhiMax;
-    };
-
-    struct WheelParams {
-        WheelGeom geom;
-        CtrlParams ctrl;
-    };
-
     // Constructor
-    UndercarriageCtrl(const std::vector<WheelParams> &params);
+    template<typename T2> UndercarriageCtrlBase(const std::vector<T2> &params){
+        for(typename std::vector<T2>::const_iterator it = params.begin(); it != params.end(); ++it){
+            wheels_.push_back(boost::make_shared<T>(*it));
+        }
+    }
 
     // Get result of direct kinematics
-    virtual void calcDirect(PlatformState &state) const;
+    virtual void calcDirect(PlatformState &state) const {
+        UndercarriageGeomBase::calcDirect(state, wheels_);
+    }
 
     // Set actual values of wheels (steer/drive velocity/position) (Istwerte)
-    virtual void updateWheelStates(const std::vector<WheelState> &states);
-
+    virtual void updateWheelStates(const std::vector<WheelState> &states) {
+        UndercarriageGeomBase::updateWheelStates(wheels_, states);
+    }
     // Set desired value for Plattform Velocity to UndercarriageCtrl (Sollwertvorgabe)
-    void setTarget(const PlatformState &state);
+    void setTarget(const PlatformState &state)  {
+        for(size_t i = 0; i < wheels_.size(); ++i){
+            wheels_[i]->setTarget(state);
+        }
+    }
 
     // Get set point values for the Wheels (including controller) from UndercarriangeCtrl
-    void calcControlStep(std::vector<WheelCommand> &commands, double dCmdRateS, bool reset);
+    void calcControlStep(std::vector<WheelCommand> &commands, double dCmdRateS, bool reset) {
+        commands.resize(wheels_.size());
 
-    void reset();
-
-    static double limitValue(double value, double limit);
-
-private:
-    struct CtrlData : public WheelData {
-        CtrlParams params_;
-
-        double m_dAngGearSteerTargetRad; // choosen alternativ for steering angle
-        double m_dVelGearDriveTargetRadS;
-
-        // previous Commanded deltaPhi e(k-1)
-        // double m_dCtrlDeltaPhi; not used
-        // previous Commanded Velocity u(k-1)
-        double m_dCtrlVelCmdInt;
-
-        // calculate inverse kinematics
-        void setTarget(const PlatformState &state);
-
-        void calcControlStep(WheelCommand &command, double dCmdRateS, bool reset);
-
-        void reset();
-
-        CtrlData(const WheelParams &params)
-        : WheelData(params.geom), params_(params.ctrl), /*m_dCtrlDeltaPhi(0),*/  m_dCtrlVelCmdInt(0) {
-            state_.dAngGearSteerRad = params_.dWheelNeutralPos;
-            updateState(WheelState());
-            setTarget(PlatformState());
-            m_dAngGearSteerTargetRad = params_.dWheelNeutralPos;
+        for(size_t i = 0; i < wheels_.size(); ++i){
+            wheels_[i]->calcControlStep(commands[i], dCmdRateS, reset);
         }
-    };
+    }
 
-    std::vector<CtrlData> wheels_;
+    void reset() {
+        for(size_t i = 0; i < wheels_.size(); ++i){
+            wheels_[i]->reset();
+        }
+    }
+
+protected:
+    std::vector<boost::shared_ptr<T> > wheels_;
+};
+
+struct CtrlParams{
+    double dWheelNeutralPos;
+
+    double dMaxDriveRateRadpS;
+    double dMaxSteerRateRadpS;
+};
+
+struct WheelCtrlParams {
+    WheelGeom geom;
+    CtrlParams ctrl;
+};
+
+struct CtrlData : public WheelData {
+    CtrlParams params_;
+
+    double m_dAngGearSteerTargetRad; // choosen alternativ for steering angle
+    double m_dVelGearDriveTargetRadS;
+
+    // calculate inverse kinematics
+    void setTarget(const PlatformState &state);
+
+    virtual void calcControlStep(WheelCommand &command, double dCmdRateS, bool reset);
+
+    virtual void reset();
+
+    template<typename P> CtrlData(const P &params)
+    : WheelData(params.geom), params_(params.ctrl) {
+        state_.dAngGearSteerRad = params_.dWheelNeutralPos;
+        updateState(WheelState());
+        setTarget(PlatformState());
+        m_dAngGearSteerTargetRad = params_.dWheelNeutralPos;
+    }
+};
+
+class UndercarriageDirectCtrl : public UndercarriageCtrlBase < CtrlData > {
+public:
+    typedef WheelCtrlParams WheelParams;
+    UndercarriageDirectCtrl(const std::vector<WheelParams> &params) : UndercarriageCtrlBase(params) {}
+};
+
+struct PosCtrlParams {
+    /** ------- Position Controller Steer Wheels -------
+    * Impedance-Ctrlr Prms
+    *  -> model Stiffness via Spring-Damper-Modell
+    *  -> only oriented at impedance-ctrl (no forces commanded)
+    *  m_dSpring   Spring-constant (elasticity)
+    *  m_dDamp             Damping coefficient (also prop. for Velocity Feedforward)
+    *  m_dVirtM    Virtual Mass of Spring-Damper System
+    *  m_dDPhiMax  maximum angular velocity (cut-off)
+    *  m_dDDPhiMax maximum angular acceleration (cut-off)
+    */
+    double dSpring, dDamp, dVirtM, dDPhiMax, dDDPhiMax;
+};
+
+struct WheelCtrlPosParams {
+    WheelGeom geom;
+    CtrlParams ctrl;
+    PosCtrlParams pos_ctrl;
+};
+
+struct PosCtrlData : public CtrlData {
+    PosCtrlParams pos_params_;
+
+    // previous Commanded deltaPhi e(k-1)
+    // double m_dCtrlDeltaPhi; not used
+    // previous Commanded Velocity u(k-1)
+    double m_dCtrlVelCmdInt;
+
+    virtual void calcControlStep(WheelCommand &command, double dCmdRateS, bool reset);
+
+    virtual void reset();
+
+    PosCtrlData(const WheelCtrlPosParams &params)
+    : CtrlData(params), pos_params_(params.pos_ctrl),m_dCtrlVelCmdInt(0) {
+        state_.dAngGearSteerRad = params_.dWheelNeutralPos;
+        updateState(WheelState());
+        setTarget(PlatformState());
+        m_dAngGearSteerTargetRad = params_.dWheelNeutralPos;
+    }
+};
+
+class UndercarriageCtrl : public UndercarriageCtrlBase < PosCtrlData > {
+public:
+    typedef WheelCtrlPosParams WheelParams;
+    UndercarriageCtrl(const std::vector<WheelParams> &params) : UndercarriageCtrlBase(params) {}
 };
 
 #endif

--- a/cob_omni_drive_controller/include/cob_omni_drive_controller/UndercarriageCtrlGeom.h
+++ b/cob_omni_drive_controller/include/cob_omni_drive_controller/UndercarriageCtrlGeom.h
@@ -331,6 +331,7 @@ struct PosCtrlData : public CtrlData {
 class UndercarriageCtrl : public UndercarriageCtrlBase < PosCtrlData > {
 public:
     typedef WheelCtrlPosParams WheelParams;
+    void configure(const std::vector<PosCtrlParams> &pos_ctrl);
     UndercarriageCtrl(const std::vector<WheelParams> &params) : UndercarriageCtrlBase(params) {}
 };
 

--- a/cob_omni_drive_controller/include/cob_omni_drive_controller/UndercarriageCtrlGeomROS.h
+++ b/cob_omni_drive_controller/include/cob_omni_drive_controller/UndercarriageCtrlGeomROS.h
@@ -8,6 +8,7 @@
 namespace cob_omni_drive_controller{
 
 bool parseWheelParams(std::vector<UndercarriageGeom::WheelParams> &params, const ros::NodeHandle &nh, bool read_urdf = true);
+bool parseWheelParams(std::vector<UndercarriageDirectCtrl::WheelParams> &params, const ros::NodeHandle &nh, bool read_urdf = true);
 bool parseWheelParams(std::vector<UndercarriageCtrl::WheelParams> &params, const ros::NodeHandle &nh, bool read_urdf = true);
 
 }

--- a/cob_omni_drive_controller/package.xml
+++ b/cob_omni_drive_controller/package.xml
@@ -18,6 +18,7 @@
   <depend>angles</depend>
   <depend>boost</depend>
   <depend>controller_interface</depend>
+  <depend>dynamic_reconfigure</depend>
   <depend>geometry_msgs</depend>
   <depend>hardware_interface</depend>
   <depend>nav_msgs</depend>

--- a/cob_omni_drive_controller/src/GeomController.h
+++ b/cob_omni_drive_controller/src/GeomController.h
@@ -3,42 +3,20 @@
 
 #include <controller_interface/controller.h>
 #include <cob_omni_drive_controller/UndercarriageCtrlGeomROS.h>
+#include <boost/scoped_ptr.hpp>
 
 namespace cob_omni_drive_controller
 {
 
-template<typename Interface, typename Controller> class GeomController: public controller_interface::Controller<Interface> {
+template<typename HandleType, typename Controller> class GeomControllerBase {
 protected:
-    std::vector<typename Interface::ResourceHandleType> steer_joints_, drive_joints_;
-    std::vector<typename Controller::WheelState> wheel_states_;
+    std::vector<HandleType> steer_joints_;
+    std::vector<HandleType> drive_joints_;
+    std::vector<WheelState> wheel_states_;
     boost::scoped_ptr<Controller> geom_;
+
 public:
-    bool init(Interface* hw, ros::NodeHandle& controller_nh){
-
-        std::vector<typename Controller::WheelParams> wheel_params;
-        if(!parseWheelParams(wheel_params, controller_nh)) return false;
-
-        if (wheel_params.size() < 3){
-            ROS_ERROR("At least three wheel are needed.");
-            return false;
-        }
-        try{
-            for (unsigned i=0; i<wheel_params.size(); i++){
-                steer_joints_.push_back(hw->getHandle(wheel_params[i].geom.steer_name));
-                drive_joints_.push_back(hw->getHandle(wheel_params[i].geom.drive_name));
-            }
-        }
-        catch(const std::exception &e){
-            ROS_ERROR_STREAM("Error while attaching handles: " << e.what());
-            return false;
-        }
-
-        wheel_states_.resize(wheel_params.size());
-        geom_.reset(new Controller(wheel_params));
-        return true;
-    }
-
-    void update(){
+    void updateState(){
 
         for (unsigned i=0; i<wheel_states_.size(); i++){
             wheel_states_[i].dAngGearSteerRad = steer_joints_[i].getPosition();
@@ -47,8 +25,40 @@ public:
         }
         geom_->updateWheelStates(wheel_states_);
     }
+protected:
+    bool setup(const std::vector<typename Controller::WheelParams> &wheel_params){
+        if (wheel_params.size() < 3){
+            ROS_ERROR("At least three wheel are needed.");
+            return false;
+        }
+        wheel_states_.resize(wheel_params.size());
+        geom_.reset(new Controller(wheel_params));
+        return true;
+
+    }
 };
 
+template<typename Interface, typename Controller> class GeomController:
+    public GeomControllerBase<typename Interface::ResourceHandleType, Controller>,
+        public controller_interface::Controller<Interface> {
+public:
+    bool init(Interface* hw, ros::NodeHandle& controller_nh){
+        std::vector<typename Controller::WheelParams> wheel_params;
+        if(!parseWheelParams(wheel_params, controller_nh)) return false;
+        if(!this->setup(wheel_params)) return false;
+        try{
+            for (unsigned i=0; i<wheel_params.size(); i++){
+                this->steer_joints_.push_back(hw->getHandle(wheel_params[i].geom.steer_name));
+                this->drive_joints_.push_back(hw->getHandle(wheel_params[i].geom.drive_name));
+            }
+        }
+        catch(const std::exception &e){
+            ROS_ERROR_STREAM("Error while attaching handles: " << e.what());
+            return false;
+        }
+        return true;
+    }
+};
 
 }
 

--- a/cob_omni_drive_controller/src/GeomController.h
+++ b/cob_omni_drive_controller/src/GeomController.h
@@ -42,9 +42,13 @@ template<typename Interface, typename Controller> class GeomController:
     public GeomControllerBase<typename Interface::ResourceHandleType, Controller>,
         public controller_interface::Controller<Interface> {
 public:
+    typedef std::vector<typename Controller::WheelParams> wheel_params_t;
     bool init(Interface* hw, ros::NodeHandle& controller_nh){
         std::vector<typename Controller::WheelParams> wheel_params;
         if(!parseWheelParams(wheel_params, controller_nh)) return false;
+        return init(hw, wheel_params);
+    }
+    bool init(Interface* hw, const wheel_params_t & wheel_params){
         if(!this->setup(wheel_params)) return false;
         try{
             for (unsigned i=0; i<wheel_params.size(); i++){

--- a/cob_omni_drive_controller/src/UndercarriageCtrlGeom.cpp
+++ b/cob_omni_drive_controller/src/UndercarriageCtrlGeom.cpp
@@ -244,3 +244,9 @@ void PosCtrlData::reset(){
     m_dCtrlVelCmdInt = 0.0;
 }
 
+void UndercarriageCtrl::configure(const std::vector<PosCtrlParams> &pos_ctrls){
+    assert(wheels_.size() == pos_ctrls.size());
+    for(size_t i = 0; i < wheels_.size(); ++i){
+        wheels_[i]->pos_params_ = pos_ctrls[i];
+    }
+}

--- a/cob_omni_drive_controller/src/UndercarriageCtrlGeom.cpp
+++ b/cob_omni_drive_controller/src/UndercarriageCtrlGeom.cpp
@@ -56,6 +56,8 @@
 #include <math.h>
 #include <angles/angles.h>
 #include <stdexcept>
+#include <boost/shared_ptr.hpp>
+
 
 namespace MathSup {
     const double PI = 3.14159265358979323846;
@@ -84,7 +86,7 @@ double getWeightedDelta(double current_position, double old_target, double new_t
     return 0.6*fabs(dtempDeltaPhi1RAD) + 0.4*fabs(dtempDeltaPhiCmd1RAD);
 }
 
-void UndercarriageGeomBase::WheelData::updateState(const WheelState &state){
+void WheelData::updateState(const WheelState &state){
     state_ = state;
 
     // calculate current geometry of robot (exact wheel position, taking into account steering offset of wheels)
@@ -100,7 +102,7 @@ void UndercarriageGeomBase::WheelData::updateState(const WheelState &state){
     m_dVelWheelMMS = geom_.dRadiusWheelMM * (state_.dVelGearDriveRadS - dFactorVel* state_.dVelGearSteerRadS);
 }
 
-double UndercarriageGeomBase::WheelData::mergeRotRobRadS(const WheelData &wheel1, const WheelData &wheel2){
+double WheelData::mergeRotRobRadS(const WheelData &wheel1, const WheelData &wheel2){
     // calc Parameters (Dist,Phi) of virtual linking axis of the two considered wheels
     double dtempDiffXMM = wheel2.m_dExWheelXPosMM - wheel1.m_dExWheelXPosMM;
     double dtempDiffYMM = wheel2.m_dExWheelYPosMM - wheel1.m_dExWheelYPosMM;
@@ -115,15 +117,15 @@ double UndercarriageGeomBase::WheelData::mergeRotRobRadS(const WheelData &wheel1
     return (wheel2.m_dVelWheelMMS * sin(dtempRelPhiWheel2RAD) - wheel1.m_dVelWheelMMS * sin(dtempRelPhiWheel1RAD))/dtempRelDistWheelsMM;
 }
 
-double UndercarriageGeomBase::WheelData::getVelX() const {
+double WheelData::getVelX() const {
     return m_dVelWheelMMS*cos(state_.dAngGearSteerRad);
 }
-double UndercarriageGeomBase::WheelData::getVelY() const {
+double WheelData::getVelY() const {
     return m_dVelWheelMMS*sin(state_.dAngGearSteerRad);
 }
 UndercarriageGeom::UndercarriageGeom(const std::vector<WheelParams> &params){
     for(std::vector<WheelParams>::const_iterator it = params.begin(); it != params.end(); ++it){
-        wheels_.push_back(WheelData(it->geom));
+        wheels_.push_back(boost::make_shared<WheelData>(it->geom));
     }
 }
 
@@ -135,7 +137,7 @@ void UndercarriageGeom::updateWheelStates(const std::vector<WheelState> &states)
     UndercarriageGeomBase::updateWheelStates(wheels_, states);
 }
 
-void UndercarriageCtrl::CtrlData::setTarget(const PlatformState &plt_state){
+void CtrlData::setTarget(const PlatformState &plt_state){
     // check if zero movement commanded -> keep orientation of wheels, set wheel velocity to zero
     if((plt_state.dVelLongMMS == 0) && (plt_state.dVelLatMMS == 0) && (plt_state.dRotRobRadS == 0))
     {
@@ -182,7 +184,7 @@ void UndercarriageCtrl::CtrlData::setTarget(const PlatformState &plt_state){
 
 }
 
-double UndercarriageCtrl::limitValue(double value, double limit){
+double limitValue(double value, double limit){
     if(limit != 0){
         if (value > limit){
                 value = limit;
@@ -193,7 +195,7 @@ double UndercarriageCtrl::limitValue(double value, double limit){
     return value;
 }
 
-void UndercarriageCtrl::CtrlData::calcControlStep(WheelCommand &command, double dCmdRateS, bool reset){
+void CtrlData::calcControlStep(WheelCommand &command, double dCmdRateS, bool reset){
     if(reset){
         this->reset();
         command.dVelGearDriveRadS = 0.0;
@@ -207,69 +209,38 @@ void UndercarriageCtrl::CtrlData::calcControlStep(WheelCommand &command, double 
     double dCurrentPosWheelRAD = angles::normalize_angle(state_.dAngGearSteerRad);
     command.dAngGearSteerRadDelta = angles::normalize_angle(m_dAngGearSteerTargetRad - dCurrentPosWheelRAD);
 
-    // Impedance-Ctrl
-    // Calculate resulting desired forces, velocities
-    // double dForceDamp, dForceProp, dAccCmd, dVelCmdInt;
-    double dForceDamp = - params_.dDamp *m_dCtrlVelCmdInt;
-    double dForceProp = params_.dSpring * command.dAngGearSteerRadDelta;
-
-    double dAccCmd = (dForceDamp + dForceProp) /  params_.dVirtM;
-    dAccCmd = limitValue(dAccCmd, params_.dDDPhiMax);
-
-    double dVelCmdInt =m_dCtrlVelCmdInt + dCmdRateS * dAccCmd;
-    dVelCmdInt = limitValue(dVelCmdInt, params_.dDPhiMax);
-
-    // Store internal ctrlr-states
-    m_dCtrlVelCmdInt = dVelCmdInt;
-
     // set outputs
-    command.dVelGearSteerRadS = limitValue(dVelCmdInt, params_.dMaxSteerRateRadpS);
     command.dVelGearDriveRadS = limitValue(m_dVelGearDriveTargetRadS + m_dAngGearSteerTargetRad * dFactorVel, params_.dMaxDriveRateRadpS);
 
     // provisorial --> skip interpolation and always take Target
     command.dAngGearSteerRad = m_dAngGearSteerTargetRad;
 }
 
-void UndercarriageCtrl::CtrlData::reset(){
-    //m_dCtrlDeltaPhi = 0.0;
+void PosCtrlData::calcControlStep(WheelCommand &command, double dCmdRateS, bool reset){
+    CtrlData::calcControlStep(command, dCmdRateS, reset);
+    if(!reset){
+        // Impedance-Ctrl
+        // Calculate resulting desired forces, velocities
+        // double dForceDamp, dForceProp, dAccCmd, dVelCmdInt;
+        double dForceDamp = - pos_params_.dDamp *m_dCtrlVelCmdInt;
+        double dForceProp = pos_params_.dSpring * command.dAngGearSteerRadDelta;
+
+        double dAccCmd = (dForceDamp + dForceProp) /  pos_params_.dVirtM;
+        dAccCmd = limitValue(dAccCmd, pos_params_.dDDPhiMax);
+
+        double dVelCmdInt =m_dCtrlVelCmdInt + dCmdRateS * dAccCmd;
+        dVelCmdInt = limitValue(dVelCmdInt, pos_params_.dDPhiMax);
+
+        // Store internal ctrlr-states
+        m_dCtrlVelCmdInt = dVelCmdInt;
+
+        // set outputs
+        command.dVelGearSteerRadS = limitValue(dVelCmdInt, params_.dMaxSteerRateRadpS);
+    }
+}
+void CtrlData::reset(){
+}
+void PosCtrlData::reset(){
     m_dCtrlVelCmdInt = 0.0;
 }
 
-UndercarriageCtrl::UndercarriageCtrl(const std::vector<WheelParams> &params){
-    for(std::vector<WheelParams>::const_iterator it = params.begin(); it != params.end(); ++it){
-        wheels_.push_back(CtrlData(*it));
-    }
-}
-
-void UndercarriageCtrl::calcDirect(PlatformState &state) const{
-    UndercarriageGeomBase::calcDirect(state, wheels_);
-}
-
-void UndercarriageCtrl::updateWheelStates(const std::vector<WheelState> &states){
-    UndercarriageGeomBase::updateWheelStates(wheels_, states);
-}
-
-void UndercarriageCtrl::setTarget(const PlatformState &state)
-{
-    for(std::vector<CtrlData>::iterator it = wheels_.begin(); it != wheels_.end(); ++it){
-        it->setTarget(state);
-    }
-}
-
-void UndercarriageCtrl::calcControlStep(std::vector<WheelCommand> &commands, double dCmdRateS, bool reset)
-{
-    commands.resize(wheels_.size());
-
-    for(size_t i = 0; i < wheels_.size(); ++i){
-        wheels_[i].calcControlStep(commands[i], dCmdRateS, reset);
-    }
-
-}
-
-void UndercarriageCtrl::reset()
-{
-    for(size_t i = 0; i < wheels_.size(); ++i){
-        wheels_[i].reset();
-    }
-
-}

--- a/cob_omni_drive_controller/src/WheelControllerBase.h
+++ b/cob_omni_drive_controller/src/WheelControllerBase.h
@@ -1,0 +1,143 @@
+#ifndef H_WHEEL_CONTROLLER_IMPL
+#define H_WHEEL_CONTROLLER_IMPL
+
+#include <controller_interface/controller.h>
+#include <hardware_interface/joint_command_interface.h>
+
+#include <pluginlib/class_list_macros.h>
+
+#include <cob_omni_drive_controller/UndercarriageCtrlGeom.h>
+#include <geometry_msgs/Twist.h>
+
+#include <boost/scoped_ptr.hpp>
+#include <boost/weak_ptr.hpp>
+#include <boost/thread/mutex.hpp>
+
+#include <realtime_tools/realtime_publisher.h>
+#include <cob_omni_drive_controller/WheelCommands.h>
+
+namespace cob_omni_drive_controller
+{
+
+template<typename T> class WheelControllerBase: public T
+{
+public:
+    bool setup(ros::NodeHandle &root_nh, ros::NodeHandle& controller_nh){
+        controller_nh.param("max_rot_velocity", max_vel_rot_, 0.0);
+        if(max_vel_rot_ < 0){
+            ROS_ERROR_STREAM("max_rot_velocity must be non-negative.");
+            return false;
+        }
+        controller_nh.param("max_trans_velocity", max_vel_trans_, 0.0);
+        if(max_vel_trans_ < 0){
+            ROS_ERROR_STREAM("max_trans_velocity must be non-negative.");
+            return false;
+        }
+        double timeout;
+        controller_nh.param("timeout", timeout, 1.0);
+        if(timeout < 0){
+            ROS_ERROR_STREAM("timeout must be non-negative.");
+            return false;
+        }
+        timeout_.fromSec(timeout);
+        
+        pub_divider_ =  controller_nh.param("pub_divider",0);
+
+        wheel_commands_.resize(this->wheel_states_.size());
+        twist_subscriber_ = controller_nh.subscribe("command", 1, &WheelControllerBase::topicCallbackTwistCmd, this);
+
+        commands_pub_.reset(new realtime_tools::RealtimePublisher<cob_omni_drive_controller::WheelCommands>(controller_nh, "wheel_commands", 5));
+       
+        commands_pub_->msg_.drive_target_velocity.resize(this->wheel_states_.size());
+        commands_pub_->msg_.steer_target_velocity.resize(this->wheel_states_.size());
+        commands_pub_->msg_.steer_target_position.resize(this->wheel_states_.size());
+        commands_pub_->msg_.steer_target_error.resize(this->wheel_states_.size());
+
+        return true;
+  }
+    virtual void starting(const ros::Time& time){
+        this->geom_->reset();
+        target_.updated = false;
+        cycles_ = 0;
+    }
+    void updateCtrl(const ros::Time& time, const ros::Duration& period){
+        {
+            boost::mutex::scoped_try_lock lock(mutex_);
+            if(lock){
+                Target target = target_;
+                target_.updated = false;
+
+                if(!target.stamp.isZero() && !timeout_.isZero() && (time - target.stamp) > timeout_){
+                    target_.stamp = ros::Time(); // only reset once
+                    target.state  = PlatformState();
+                    target.updated = true;
+                }
+                lock.unlock();
+
+                if(target.updated){
+                   this->geom_->setTarget(target.state);
+                }
+            }
+        }
+
+        this->geom_->calcControlStep(wheel_commands_, period.toSec(), false);
+
+        if(cycles_ < pub_divider_ && (++cycles_) == pub_divider_){
+            if(commands_pub_->trylock()){
+                ++(commands_pub_->msg_.header.seq);
+                commands_pub_->msg_.header.stamp = time;
+
+                for (unsigned i=0; i<wheel_commands_.size(); i++){
+                    commands_pub_->msg_.drive_target_velocity[i] = wheel_commands_[i].dVelGearDriveRadS;
+                    commands_pub_->msg_.steer_target_velocity[i] = wheel_commands_[i].dVelGearSteerRadS;
+                    commands_pub_->msg_.steer_target_position[i] = wheel_commands_[i].dAngGearSteerRad;
+                    commands_pub_->msg_.steer_target_error[i] = wheel_commands_[i].dAngGearSteerRadDelta;
+                }
+                commands_pub_->unlockAndPublish();
+            
+            }
+            cycles_ = 0;
+        }
+    }
+    virtual void stopping(const ros::Time& time) {}
+
+protected:
+    struct Target {
+        PlatformState state;
+        bool updated;
+        ros::Time stamp;
+    } target_;
+
+    std::vector<WheelCommand> wheel_commands_;
+
+    boost::mutex mutex_;
+    ros::Subscriber twist_subscriber_;
+    
+    boost::scoped_ptr<realtime_tools::RealtimePublisher<cob_omni_drive_controller::WheelCommands> > commands_pub_;
+    uint32_t cycles_;
+    uint32_t pub_divider_;
+    
+    ros::Duration timeout_;
+    double max_vel_trans_, max_vel_rot_;
+
+    void topicCallbackTwistCmd(const geometry_msgs::Twist::ConstPtr& msg){
+        if(this->isRunning()){
+            boost::mutex::scoped_lock lock(mutex_);
+            if(isnan(msg->linear.x) || isnan(msg->linear.y) || isnan(msg->angular.z)) {
+                ROS_FATAL("Received NaN-value in Twist message. Reset target to zero.");
+                target_.state = PlatformState();
+            }else{
+                target_.state.setVelX(limitValue(msg->linear.x, max_vel_trans_));
+                target_.state.setVelY(limitValue(msg->linear.y, max_vel_trans_));
+                target_.state.dRotRobRadS = limitValue(msg->angular.z, max_vel_rot_);
+            }
+            target_.updated = true;
+            target_.stamp = ros::Time::now();
+        }
+    }
+
+
+};
+
+}
+#endif

--- a/cob_omni_drive_controller/src/control_plugin.cpp
+++ b/cob_omni_drive_controller/src/control_plugin.cpp
@@ -1,154 +1,30 @@
-
-#include <controller_interface/controller.h>
-#include <hardware_interface/joint_command_interface.h>
-
-#include <pluginlib/class_list_macros.h>
-
-#include <cob_omni_drive_controller/UndercarriageCtrlGeom.h>
-#include <geometry_msgs/Twist.h>
-
-#include <boost/scoped_ptr.hpp>
-#include <boost/weak_ptr.hpp>
-#include <boost/thread/mutex.hpp>
-
 #include "GeomController.h"
-
-#include <realtime_tools/realtime_publisher.h>
-#include <cob_omni_drive_controller/WheelCommands.h>
+#include "WheelControllerBase.h"
 
 namespace cob_omni_drive_controller
 {
 
-class WheelController: public GeomController<hardware_interface::VelocityJointInterface, UndercarriageCtrl>
+class WheelController : public WheelControllerBase< GeomController<hardware_interface::VelocityJointInterface, UndercarriageCtrl> >
 {
 public:
-    WheelController() {}
-
     virtual bool init(hardware_interface::VelocityJointInterface* hw, ros::NodeHandle &root_nh, ros::NodeHandle& controller_nh){
 
         if(!GeomController::init(hw, controller_nh)) return false;
 
-        controller_nh.param("max_rot_velocity", max_vel_rot_, 0.0);
-        if(max_vel_rot_ < 0){
-            ROS_ERROR_STREAM("max_rot_velocity must be non-negative.");
-            return false;
-        }
-        controller_nh.param("max_trans_velocity", max_vel_trans_, 0.0);
-        if(max_vel_trans_ < 0){
-            ROS_ERROR_STREAM("max_trans_velocity must be non-negative.");
-            return false;
-        }
-        double timeout;
-        controller_nh.param("timeout", timeout, 1.0);
-        if(timeout < 0){
-            ROS_ERROR_STREAM("timeout must be non-negative.");
-            return false;
-        }
-        timeout_.fromSec(timeout);
-        
-        pub_divider_ =  controller_nh.param("pub_divider",0);
-
-        wheel_commands_.resize(wheel_states_.size());
-        twist_subscriber_ = controller_nh.subscribe("command", 1, &WheelController::topicCallbackTwistCmd, this);
-
-        commands_pub_.reset(new realtime_tools::RealtimePublisher<cob_omni_drive_controller::WheelCommands>(controller_nh, "wheel_commands", 5));
-       
-        commands_pub_->msg_.drive_target_velocity.resize(wheel_states_.size());
-        commands_pub_->msg_.steer_target_velocity.resize(wheel_states_.size());
-        commands_pub_->msg_.steer_target_position.resize(wheel_states_.size());
-        commands_pub_->msg_.steer_target_error.resize(wheel_states_.size());
-
-        return true;
-  }
-    virtual void starting(const ros::Time& time){
-        geom_->reset();
-        target_.updated = false;
-        cycles_ = 0;
+        return setup(root_nh,controller_nh);
     }
     virtual void update(const ros::Time& time, const ros::Duration& period){
 
-        GeomController::update();
+        updateState();
 
-        {
-            boost::mutex::scoped_try_lock lock(mutex_);
-            if(lock){
-                Target target = target_;
-                target_.updated = false;
-
-                if(!target.stamp.isZero() && !timeout_.isZero() && (time - target.stamp) > timeout_){
-                    target_.stamp = ros::Time(); // only reset once
-                    target.state  = UndercarriageCtrl::PlatformState();
-                    target.updated = true;
-                }
-                lock.unlock();
-
-                if(target.updated){
-                   geom_->setTarget(target.state);
-                }
-            }
-        }
-
-        geom_->calcControlStep(wheel_commands_, period.toSec(), false);
+        updateCtrl(time, period);
 
         for (unsigned i=0; i<wheel_commands_.size(); i++){
             steer_joints_[i].setCommand(wheel_commands_[i].dVelGearSteerRadS);
             drive_joints_[i].setCommand(wheel_commands_[i].dVelGearDriveRadS);
         }
-        
-        if(cycles_ < pub_divider_ && (++cycles_) == pub_divider_){
-            if(commands_pub_->trylock()){
-                ++(commands_pub_->msg_.header.seq);
-                commands_pub_->msg_.header.stamp = time;
 
-                for (unsigned i=0; i<wheel_commands_.size(); i++){
-                    commands_pub_->msg_.drive_target_velocity[i] = wheel_commands_[i].dVelGearDriveRadS;
-                    commands_pub_->msg_.steer_target_velocity[i] = wheel_commands_[i].dVelGearSteerRadS;
-                    commands_pub_->msg_.steer_target_position[i] = wheel_commands_[i].dAngGearSteerRad;
-                    commands_pub_->msg_.steer_target_error[i] = wheel_commands_[i].dAngGearSteerRadDelta;
-                }
-                commands_pub_->unlockAndPublish();
-            
-            }
-            cycles_ = 0;
-        }
     }
-    virtual void stopping(const ros::Time& time) {}
-
-private:
-    struct Target {
-        UndercarriageCtrl::PlatformState state;
-        bool updated;
-        ros::Time stamp;
-    } target_;
-
-    std::vector<UndercarriageCtrl::WheelCommand> wheel_commands_;
-
-    boost::mutex mutex_;
-    ros::Subscriber twist_subscriber_;
-    
-    boost::scoped_ptr<realtime_tools::RealtimePublisher<cob_omni_drive_controller::WheelCommands> > commands_pub_;
-    uint32_t cycles_;
-    uint32_t pub_divider_;
-    
-    ros::Duration timeout_;
-    double max_vel_trans_, max_vel_rot_;
-
-    void topicCallbackTwistCmd(const geometry_msgs::Twist::ConstPtr& msg){
-        if(isRunning()){
-            boost::mutex::scoped_lock lock(mutex_);
-            if(isnan(msg->linear.x) || isnan(msg->linear.y) || isnan(msg->angular.z)) {
-                ROS_FATAL("Received NaN-value in Twist message. Reset target to zero.");
-                target_.state = UndercarriageCtrl::PlatformState();
-            }else{
-                target_.state.setVelX(UndercarriageCtrl::limitValue(msg->linear.x, max_vel_trans_));
-                target_.state.setVelY(UndercarriageCtrl::limitValue(msg->linear.y, max_vel_trans_));
-                target_.state.dRotRobRadS = UndercarriageCtrl::limitValue(msg->angular.z, max_vel_rot_);
-            }
-            target_.updated = true;
-            target_.stamp = ros::Time::now();
-        }
-    }
-
 
 };
 

--- a/cob_omni_drive_controller/src/control_plugin.cpp
+++ b/cob_omni_drive_controller/src/control_plugin.cpp
@@ -1,6 +1,9 @@
 #include "GeomController.h"
 #include "WheelControllerBase.h"
 
+#include <dynamic_reconfigure/server.h>
+#include <cob_omni_drive_controller/SteerCtrlConfig.h>
+
 namespace cob_omni_drive_controller
 {
 
@@ -9,13 +12,17 @@ class WheelController : public WheelControllerBase< GeomController<hardware_inte
 public:
     virtual bool init(hardware_interface::VelocityJointInterface* hw, ros::NodeHandle &root_nh, ros::NodeHandle& controller_nh){
 
-        if(!GeomController::init(hw, controller_nh)) return false;
+        wheel_params_t wheel_params;
+        if(!parseWheelParams(wheel_params, controller_nh) || !GeomController::init(hw, wheel_params)) return false;
 
-        return setup(root_nh,controller_nh);
+        pos_ctrl_.init(wheel_params, controller_nh);
+
+        return setup(root_nh, controller_nh);
     }
     virtual void update(const ros::Time& time, const ros::Duration& period){
 
         updateState();
+        pos_ctrl_.try_configure(*geom_);
 
         updateCtrl(time, period);
 
@@ -25,6 +32,77 @@ public:
         }
 
     }
+    class PosCtrl {
+    public:
+        PosCtrl() : updated(false) {}
+        void try_configure(UndercarriageCtrl &ctrl){
+            boost::recursive_mutex::scoped_try_lock lock(mutex);
+            if(lock && updated){
+                ctrl.configure(pos_ctrl_params);
+                updated = false;
+            }
+        }
+        void init(const wheel_params_t &params, const ros::NodeHandle &nh){
+            boost::recursive_mutex::scoped_lock lock(mutex);
+            pos_ctrl_params.resize(params.size());
+            reconfigure_server_axes_.clear();
+
+            reconfigure_server_.reset( new dynamic_reconfigure::Server<SteerCtrlConfig>(mutex,ros::NodeHandle(nh, "default/steer_ctrl")));
+            reconfigure_server_->setCallback(boost::bind(&PosCtrl::setForAll, this, _1, _2)); // this writes the default into pos_ctrl_params
+            {
+                SteerCtrlConfig config;
+                copy(config, params.front().pos_ctrl);
+                reconfigure_server_->setConfigDefault(config);
+            }
+
+            for(size_t i=0; i< pos_ctrl_params.size(); ++i) {
+                boost::shared_ptr<dynamic_reconfigure::Server<SteerCtrlConfig> > dr(new dynamic_reconfigure::Server<SteerCtrlConfig>(mutex, ros::NodeHandle(nh, params[i].geom.steer_name)));
+                SteerCtrlConfig config;
+                copy(config, params[i].pos_ctrl);
+                dr->setConfigDefault(config);
+                dr->updateConfig(config);
+                dr->setCallback(boost::bind(&PosCtrl::setForOne, this, i, _1, _2));  // this writes the joint-specific config into pos_ctrl_params
+                reconfigure_server_axes_.push_back(dr);
+            }
+        }
+    private:
+        static void copy(PosCtrlParams &params, const SteerCtrlConfig &config){
+            params.dSpring = config.spring;
+            params.dDamp = config.damp;
+            params.dVirtM = config.virt_mass;
+            params.dDPhiMax = config.d_phi_max;
+            params.dDDPhiMax = config.dd_phi_max;
+        }
+        static void copy(SteerCtrlConfig &config, const PosCtrlParams &params){
+            config.spring = params.dSpring;
+            config.damp = params.dDamp;
+            config.virt_mass = params.dVirtM;
+            config.d_phi_max = params.dDPhiMax;
+            config.dd_phi_max = params.dDDPhiMax;
+        }
+        // these function don't get locked
+        void setForAll(SteerCtrlConfig &config, uint32_t /*level*/) {
+            ROS_INFO("configure all steers: s: %lf, d: %lf, m: %lf, v: %lf, a: %lf", config.spring, config.damp, config.virt_mass, config.d_phi_max, config.dd_phi_max);
+            for(size_t i=0; i< pos_ctrl_params.size(); ++i) {
+                copy(pos_ctrl_params[i], config);
+                if(!reconfigure_server_axes_.empty()){
+                    reconfigure_server_axes_[i]->setConfigDefault(config);
+                    reconfigure_server_axes_[i]->updateConfig(config);
+                }
+            }
+            updated = true;
+        }
+        void setForOne(size_t i, SteerCtrlConfig &config, uint32_t /*level*/) {
+            ROS_INFO("configure steer %d: s: %lf, d: %lf, m: %lf, v: %lf, a: %lf", (int)i, config.spring, config.damp, config.virt_mass, config.d_phi_max, config.dd_phi_max);
+            copy(pos_ctrl_params[i], config);
+            updated = true;
+        }
+        std::vector<PosCtrlParams> pos_ctrl_params;
+        boost::recursive_mutex mutex; // dynamic_reconfigure::Server calls the callback from the setCallback function
+        bool updated;
+        boost::scoped_ptr< dynamic_reconfigure::Server<SteerCtrlConfig> > reconfigure_server_;
+        std::vector<boost::shared_ptr< dynamic_reconfigure::Server<SteerCtrlConfig> > > reconfigure_server_axes_; // TODO: use unique_ptr
+    } pos_ctrl_;
 
 };
 

--- a/cob_omni_drive_controller/src/odom_plugin.cpp
+++ b/cob_omni_drive_controller/src/odom_plugin.cpp
@@ -82,7 +82,7 @@ public:
 
     virtual void update(const ros::Time& time, const ros::Duration& period){
 
-        GeomController::update();
+        updateState();
 
         geom_->calcDirect(platform_state_);
 
@@ -101,7 +101,7 @@ public:
     virtual void stopping(const ros::Time& time) { stop_time_ = time; }
 
 private:
-    UndercarriageGeom::PlatformState platform_state_;
+    PlatformState platform_state_;
 
     ros::Publisher topic_pub_odometry_;                 // calculated (measured) velocity, rotation and pose (odometry-based) for the robot
     ros::ServiceServer service_reset_;                  // service to reset odometry to zero

--- a/cob_omni_drive_controller/src/param_parser.cpp
+++ b/cob_omni_drive_controller/src/param_parser.cpp
@@ -58,14 +58,18 @@ template<typename T> bool read(T& val, const std::string &name, XmlRpc::XmlRpcVa
     return true;
 }
 
-bool parseCtrlParams(UndercarriageCtrl::CtrlParams & params, XmlRpc::XmlRpcValue &wheel){
+bool parseCtrlParams(CtrlParams & params, XmlRpc::XmlRpcValue &wheel){
     double deg;
     read_with_default(deg, "steer_neutral_position", wheel, 0.0);
     params.dWheelNeutralPos = angles::from_degrees(deg);
 
     read_with_default(params.dMaxSteerRateRadpS, "max_steer_rate", wheel, 0.0);
     read_with_default(params.dMaxDriveRateRadpS, "max_drive_rate", wheel, 0.0);
+    
+    return true;
+}
 
+bool parsePosCtrlParams(PosCtrlParams & params, XmlRpc::XmlRpcValue &wheel){
     if(!wheel.hasMember("steer_ctrl")){
         ROS_ERROR_STREAM("steer_ctrl not found");
         return false;
@@ -79,7 +83,7 @@ bool parseCtrlParams(UndercarriageCtrl::CtrlParams & params, XmlRpc::XmlRpcValue
         && read(params.dDDPhiMax, "dd_phi_max", steer);
 }
 
-bool parseWheelGeom(UndercarriageGeom::WheelGeom & geom, XmlRpc::XmlRpcValue &wheel, MergedXmlRpcStruct &merged, urdf::Model* model){
+bool parseWheelGeom(WheelGeom & geom, XmlRpc::XmlRpcValue &wheel, MergedXmlRpcStruct &merged, urdf::Model* model){
 
     read_with_default(geom.steer_name, "steer", wheel, std::string());
     read_with_default(geom.drive_name, "drive", wheel, std::string());
@@ -144,8 +148,12 @@ template<> bool parseWheel(UndercarriageGeom::WheelParams & params, XmlRpc::XmlR
     return parseWheelGeom(params.geom, wheel, merged, model);
 }
 
-template<> bool parseWheel(UndercarriageCtrl::WheelParams & params, XmlRpc::XmlRpcValue &wheel, MergedXmlRpcStruct &merged, urdf::Model* model){
+template<> bool parseWheel(UndercarriageDirectCtrl::WheelParams & params, XmlRpc::XmlRpcValue &wheel, MergedXmlRpcStruct &merged, urdf::Model* model){
     return parseWheelGeom(params.geom, wheel, merged, model) && parseCtrlParams(params.ctrl, merged);
+}
+
+template<> bool parseWheel(UndercarriageCtrl::WheelParams & params, XmlRpc::XmlRpcValue &wheel, MergedXmlRpcStruct &merged, urdf::Model* model){
+    return parseWheelGeom(params.geom, wheel, merged, model) && parseCtrlParams(params.ctrl, merged) && parsePosCtrlParams(params.pos_ctrl, merged);
 }
 
 bool make_wheel_struct(XmlRpc::XmlRpcValue &wheel_list){
@@ -203,6 +211,9 @@ template<typename W> bool parseWheels(std::vector<W> &wheel_params, const ros::N
 namespace cob_omni_drive_controller{
 
 bool parseWheelParams(std::vector<UndercarriageGeom::WheelParams> &params, const ros::NodeHandle &nh, bool read_urdf /* = true*/){
+    return parseWheels(params, nh, read_urdf);
+}
+bool parseWheelParams(std::vector<UndercarriageDirectCtrl::WheelParams> &params, const ros::NodeHandle &nh, bool read_urdf /* = true*/){
     return parseWheels(params, nh, read_urdf);
 }
 bool parseWheelParams(std::vector<UndercarriageCtrl::WheelParams> &params, const ros::NodeHandle &nh, bool read_urdf /* = true*/){

--- a/cob_undercarriage_ctrl_node/src/cob_undercarriage_ctrl_new.cpp
+++ b/cob_undercarriage_ctrl_node/src/cob_undercarriage_ctrl_new.cpp
@@ -262,7 +262,7 @@ class NodeClass
     void topicCallbackTwistCmd(const geometry_msgs::Twist::ConstPtr& msg)
     {
       // create instance of pltState which is initialized with zero values
-      UndercarriageCtrl::PlatformState pltState;
+      PlatformState pltState;
 
       if( (fabs(msg->linear.x) > max_vel_trans_) || (fabs(msg->linear.y) > max_vel_trans_) || (fabs(msg->angular.z) > max_vel_rot_)){
         if(fabs(msg->linear.x) > max_vel_trans_){
@@ -304,7 +304,7 @@ class NodeClass
         //std::cout << "PLT_VELO_NEW: " << pltState.dVelLongMMS << " , " << pltState.dVelLatMMS << " , " << pltState.dRotRobRadS << std::endl;
         // Set desired values for Plattform State to zero (setpoint setting)
         // pltState will be initialized with zero values
-        pltState = UndercarriageCtrl::PlatformState();
+        pltState = PlatformState();
         ucar_ctrl_->setTarget(pltState);
 
         ROS_DEBUG("Forced platform-velocity cmds to zero");
@@ -328,7 +328,7 @@ class NodeClass
 
         // Set desired values for Plattform State to zero (setpoint setting)
         // pltState will be initialized with zero values
-        UndercarriageCtrl::PlatformState pltState;
+        PlatformState pltState;
         ucar_ctrl_->setTarget(pltState);
 
         ROS_DEBUG("Forced platform-velocity cmds to zero");
@@ -388,7 +388,7 @@ class NodeClass
 
             // Set desired values for Plattform State to zero (setpoint setting)
             // pltState will be initialized with zero values
-            UndercarriageCtrl::PlatformState pltState;
+            PlatformState pltState;
             ucar_ctrl_->setTarget(pltState);
 
             ROS_DEBUG("Forced platform-velocity cmds to zero");
@@ -410,8 +410,8 @@ class NodeClass
       int num_joints = msg->joint_names.size();
 
       // replaces the vectors per parameter with a vector of wheelStates which combines the wheel specfic params
-      std::vector<UndercarriageCtrl::WheelState> wStates;
-      wStates.assign(m_iNumWheels, UndercarriageCtrl::WheelState());
+      std::vector<WheelState> wStates;
+      wStates.assign(m_iNumWheels, WheelState());
 
       joint_state_odom_stamp_ = msg->header.stamp;
 
@@ -520,8 +520,8 @@ int main(int argc, char** argv)
 void NodeClass::CalcCtrlStep()
 {
   // WheelStates will be initialized with zero-values
-  std::vector<UndercarriageCtrl::WheelCommand> wStates;
-  wStates.assign(m_iNumWheels, UndercarriageCtrl::WheelCommand());
+  std::vector<WheelCommand> wStates;
+  wStates.assign(m_iNumWheels, WheelCommand());
 
   // create control_msg
   control_msgs::JointTrajectoryControllerState joint_state_cmd;
@@ -612,7 +612,7 @@ void NodeClass::UpdateOdometry()
 {
   // if drive chain already initialized process joint data
   //if (drive_chain_diagnostic_ != diagnostic_status_lookup_.OK)
-  UndercarriageCtrl::PlatformState pltState;
+  PlatformState pltState;
   if (is_initialized_bool_)
   {
 
@@ -652,11 +652,11 @@ void NodeClass::setEMStopActive(bool bEMStopActive)
 {
     m_bEMStopActive = bEMStopActive;
 
-    std::vector<UndercarriageCtrl::WheelState> wStates;
-    wStates.assign(m_iNumWheels, UndercarriageCtrl::WheelState());
+    std::vector<WheelState> wStates;
+    wStates.assign(m_iNumWheels, WheelState());
 
-    std::vector<UndercarriageCtrl::WheelCommand> wCommands;
-    wCommands.assign(m_iNumWheels, UndercarriageCtrl::WheelCommand());
+    std::vector<WheelCommand> wCommands;
+    wCommands.assign(m_iNumWheels, WheelCommand());
     
     // if emergency stop reset ctrlr to zero
     if(m_bEMStopActive)

